File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
from sys import stdin as s
#s = open("input.txt")

N = int(s.readline().rstrip())

inputList = list(map(int, s.readline().split(" ")))

dp = [1] * N

for i in range(1, N):
    item = inputList[i]
    for j in range(i-1, -1, -1):
        if (inputList[j] < item) and (dp[j] >= dp[i]):
            dp[i] = dp[j] + 1
print(max(dp))

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.js

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split("\n");

const N = Number(input.shift());
const arr = input.shift().split(" ").map(Number);
const ans = Array.from({ length: N }, () => 0);
ans[0] = 1;

for (let i = 1; i < N; i++) {
  let mv = 0;
  for (let j = 0; j < i; j++) {
    if (arr[i] > arr[j]) {
      mv = Math.max(ans[j], mv);
    }
  }
  ans[i] = mv + 1;
}

console.log(Math.max(...ans));
